fenicsprecice/adapter_core.py

```diff
@@ -427,34 +427,35 @@ def get_forces_as_point_sources(fixed_boundary, function_space, data):
 
     Returns
     -------
-    x_forces : list
-        Dictionary carrying X component of forces with reference to each point on the coupling interface.
-    y_forces : list
-        Dictionary carrying Y component of forces with reference to each point on the coupling interface.
+    forces : list
+        Multiple lists carrying components of forces with reference to each point on the coupling interface.
     """
-    x_forces = dict()  # dict of PointSources for Forces in x direction
-    y_forces = dict()  # dict of PointSources for Forces in y direction
-
-    fenics_vertices = np.array(list(data.keys()))
+    vertices = np.array(list(data.keys()))
     nodal_data = np.array(list(data.values()))
 
-    # Check for shape of coupling_mesh_vertices and raise Assertion for 3D
-    n_vertices, _ = fenics_vertices.shape
+    n_vertices, dims = vertices.shape
 
-    vertices_x = fenics_vertices[:, 0]
-    vertices_y = fenics_vertices[:, 1]
+    forces = []
+    for d in range(dims):
+        forces.append(dict())
 
     for i in range(n_vertices):
-        px, py = vertices_x[i], vertices_y[i]
-        key = (px, py)
-        x_forces[key] = PointSource(function_space.sub(0), Point(px, py), nodal_data[i, 0])
-        y_forces[key] = PointSource(function_space.sub(1), Point(px, py), nodal_data[i, 1])
+        key = []
+        for d in range(dims):
+            key.append(vertices[i, d])
+        key = tuple(key)
+
+        for d in range(dims):
+            forces[d][key] = PointSource(function_space.sub(d), Point(key), nodal_data[i, d])
 
     # Avoid application of PointSource and Dirichlet boundary condition at the same point by filtering
-    x_forces = filter_point_sources(x_forces, fixed_boundary, warn_duplicate=False)
-    y_forces = filter_point_sources(y_forces, fixed_boundary, warn_duplicate=False)
+    for d in range(dims):
+        forces[d] = filter_point_sources(forces[d], fixed_boundary, warn_duplicate=False)
 
-    return x_forces.values(), y_forces.values()  # don't return dictionary, but list of PointSources
+    if dims == 2:
+        return forces[0].values(), forces[1].values()  # don't return dictionary, but lists of PointSources
+    elif dims == 3:
+        return forces[0].values(), forces[1].values(), forces[2].values()
```

Comment:
Better return a `d x n` (or `n x d`?) array. I would avoid returning a different number of return values depending on the number of dimensions. This would also resolve the issue above with the documentation.

Additionally: I think this would make it more consistent with the changes introduced in #145.

Reply:
I am a bit hesitant to change this as it will affect all tutorial cases and in general affect the user-side API of the adapter. Should we still do this now?

Reply:
We should definitely try to keep the API stable. But are you sure? I thought the user only accesses [`precice.get_point_sources`](https://github.com/precice/tutorials/blob/af4282de60f56d8f684cd0caee19acd1f6d69f0f/perpendicular-flap/solid-fenics/solid.py#L187-L188). This function here is in `adapter_core` and hidden from the user. Or am I missing something?

Btw: Do you know that you can unpack return values of a functions: If you `return [first, second]` of `return a_list_of_two` you can receive the return values via `first, second = my_function_returning_a_list_of_two`? Maybe this is a good way of having a consistent return value for different dimensions and at the same time a stable API.

Reply:
The function itself is hidden from the user but we basically return the return of this internal function directly in the following way:
https://github.com/precice/fenics-adapter/blob/f517f2c27b3b2ccf0f0c90359253079ff976833e/fenicsprecice/fenicsprecice.py#L189

hence changing the return statement would inadvertently change the API.

I tried to have unpacked lists in the return statement but it did not work. I will try again.